WIP Enhance CSV Capabilities

kp2401075 opened the pull request:
Read CSVs with Empty column names
Read CSVs with Duplicate Column names.

This is to fix #34

kp2401075 commented:
@tobgu Can you please review this and suggest any corrections.

Thanks,

tobgu commented:
Thanks for the PR! I've given this some thought and have some ideas/suggestions that you may want to consider.

With the current solution we address two issues in one package (empty columns and duplicates) which is sort of fine but the user cannot control how this is done in any way (empty names are replaced with "column<colIx>" while other duplicates get the name "<name><colIx>"). The new config variable doesn't really provide a hint of how this is done. It simply states that we'll rename duplicates.

I have a suggestion: we move the details of how the rename and duplicate handling is done into a "policy function" and let this policy function be part of the `Config` object instead. The "policy function" could have the following signature:
`func(colIndex int, colName string) (newName string)` and we could set a default implementation of it that would simply be the identity function:
```go
func defaultPolicy(_ int, colName string) string {
   return colName
}
```

You can then implement more or less the code you've written in a new policy function which is constructed providing the prefix to use for empty columns as a string:

```go
// This would allow a user of the library to implement any policy they want.
func ColumnNamePolicy(func(colIndex int, colName string) string) ConfigFunc {
   // Return a function that sets policy func on config object
}

func AddColumnIndexToDuplicateColumns(emptyColNamePrefix string) ConfigFunc {
  policyFunc := func(colIx int, colName string) string {
    // Your existing code goes here
  }
  return ColumnNamePolicy(policyFunc)
}
```

In `ReadCSV` there would simply be a loop somewhere near the top to apply the policy function to each header index and name and replace the names read or passed in with those produced by the policy function.

There are obviously a few moving parts involved in the solution proposed above and there may be simpler ways to achieve the same thing. Please let me know your thoughts. Also, I realize I may have made a bad job trying to convey my ideas above. Please ask for clarifications in that case. I can provide some more code snippets for example.

Also, please add tests verifying the important parts of any new code that you add!

kp2401075 commented:
Hi @tobgu,

I think There is a simpler way to achieve this, I know I did a dumb thing and implemented 2 things in to 1.

but we can break that in to 2 config options.
1. For empty column names
2. For duplicate column names


Something like below:

and we can think of better name for config options.

```go
//For reading  CSV
type CSVConfig struct {
	EmptyNull        bool
	IgnoreEmptyLines bool
	Delimiter        byte
	Types            map[string]types.DataType
	EnumVals         map[string][]string
	RowCountHint     int
	Headers          []string
	AddColumnIndexToDuplicateColumns bool
	AddColumnIndexToEmptyColumns bool
}

```
we can give examples in Documentation for explaining what each option will do to user's Input. This should make it straight forward for user.

If you think solution I proposed is not a good fit then I can try and implement your proposal from https://github.com/tobgu/qframe/pull/39#issuecomment-841874131

Have a think and let me know.

tobgu commented:
Hi!

I think something like what you describe above is probably good enough and, as you say, simpler.

I would suggest changing `AddColumnIndexToEmptyColumns bool` to something like `MissingColumnNameAlias string` to give the caller the possibility to set the name of empty columns rather than hard coding it in the lib. The default could be "", which would lead to the current behaviour, eg. rejecting empty column names. To make this work with multiple empty column names `AddColumnIndexToDuplicateColumns` would also have to be set to true.

Code examples are always great to add and would be highly appreciated!

kp2401075 commented:
I've made appropriate changes & also added some test functions to test when building.

I've also added some Comments in the config functions to show what this option does to the user input.

when the documentation gets generated it should show up.

Can you please take another pass at it.

tobgu commented on internal/io/csv.go:
```diff
@@ -98,6 +101,37 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
+
+			if headerSet.Contains(h) {
+				h = h + fmt.Sprint(count)
+			} else {
```
I think this part potentially needs to be thought through a bit more to make it bullet proof wrt to duplicates. As is now you could construct header combinations that will fail the uniqueness constraints.
`COL,COL,COL1`, `COL,COL2,COL`

While probably not super common I'd prefer if we could handle this since its a library and you don't really know what users will use it for. We could come up with a naming scheme like <NAME><COL>.<ENUMERATOR> where the ENUMERATOR is just added in case there are conflicts and is incremented until it does not conflict with an already existing column name.

It would be great to add a couple of unit tests for these corner cases.

tobgu commented on internal/io/csv.go:
```diff
@@ -98,6 +101,37 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
+				headerSet.Add(h)
+			}
+			headers[count] = h
+
```
Remove empty line

tobgu commented on internal/io/csv.go:
```diff
@@ -98,6 +101,37 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
+			headers[count] = h
+
+		}
+		// fmt.Println("Duplicates have been renamed")
```
Remove comment

tobgu commented on internal/io/csv.go:
```diff
@@ -116,17 +150,20 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
 		duplicates := make([]string, 0)
 		headerSet := strings.NewEmptyStringSet()
 		for _, h := range headers {
+
```
Remove empty line

tobgu commented on internal/io/csv.go:
```diff
@@ -116,17 +150,20 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
 	}
 
 	return dataMap, headers, nil
+
```
Remove empty line

tobgu commented on qframe_test.go:
```diff
@@ -1373,6 +1373,56 @@ mon
+func TestQFrame_ReadCSVMissingColumnName(t *testing.T) {
+	input := `,COL2
+a,1.5
+b,2.25
```
Perhaps reduce row count to one line since only one row is needed?

tobgu commented on qframe_test.go:
```diff
@@ -1373,6 +1373,56 @@ mon
+	expected := qframe.ReadCSV(strings.NewReader(expectedIn))
+	assertEquals(t, expected, out)
+
+	// Output:
```
Remove this comment or migrate this test to be an example test. See the examples in "examples_test.go". Those are part of the documentation and adding the happy/base cases there may be a good idea.

tobgu commented on qframe_test.go:
```diff
@@ -1373,6 +1373,56 @@ mon
+	out := qframe.ReadCSV(strings.NewReader(input), csv.AddColumnIndexToDuplicateColumns(true))
+	expected := qframe.ReadCSV(strings.NewReader(expectedIn))
+	assertEquals(t, expected, out)
+	// Output:
```
See comment about examples in test above.

tobgu commented on config/csv/config.go:
```diff
@@ -33,6 +33,31 @@ func EmptyNull(emptyNull bool) ConfigFunc {
+
+// AddColumnIndexToDuplicateColumns configures if column names have duplicates
+// column index will be added to the column name
+// COL,COL 			// COL0   COL1
```
I think we can skip the inline examples here and instead create proper example tests covering the usage.

kp2401075 replied:
I need help with this, What should be the function names in examples_test.go file. for examples.

tobgu commented on config/csv/config.go:
```diff
@@ -33,6 +33,31 @@ func EmptyNull(emptyNull bool) ConfigFunc {
 }
 
+// MissingColumnNameAliasString set configs to rename empty column name with given string
+//  ,COL2 			// col   COL2
```
Skip inline example here? See similar comment below.

tobgu commented on config/csv/config.go:
```diff
@@ -33,6 +33,31 @@ func EmptyNull(emptyNull bool) ConfigFunc {
 	}
 }
 
+// MissingColumnNameAliasString set configs to rename empty column name with given string
```
"MissingColumnNameAliasString..." -> "MissingColumnNameAlias sets the name to be used for empty columns..."
Perhaps add a comment here about the fact that you probably also want to set `AddColumnIndexToDuplicateColumns` if you expect to have multiple columns with missing names?

tobgu commented on config/csv/config.go:
```diff
@@ -33,6 +33,31 @@ func EmptyNull(emptyNull bool) ConfigFunc {
+	}
+}
+
+// AddColumnIndexToDuplicateColumns configures if column names have duplicates
```
"... if duplicate column names should have the column index appended to the column name to resolve the conflict.", or similar?

tobgu commented on internal/io/csv.go:
```diff
@@ -20,13 +21,15 @@ type bytePointer struct {
+	RowCountHint                     int
+	Headers                          []string
+	AddColumnIndexToDuplicateColumns bool
+	MissingColumnNameAlias           string ""
```
What are the double quotes for?

tobgu reviewed (changes requested):
Thanks! I've dropped a couple of comments/questions/suggestions that you may want to consider.

tobgu commented on internal/io/csv.go:
```diff
@@ -98,6 +101,27 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
+	if conf.AddColumnIndexToDuplicateColumns {
+		headerSet := strings.NewEmptyStringSet()
+		for i, h := range headers {
+			if headerSet.Contains(h + fmt.Sprint(i)) {
```
What would happen for ´COL1,COL,COL11´ (which is valid as is) for example?

I think you somehow need to take into consideration not only the headers that you've looped over but also those that come later since you don't want to create conflicts with existing column names. Perhaps a ´map[string]int´ mapping the column name to the first column position of the name  that you populate initially would be a better data structure for this than the stringset?

kp2401075 replied:
I knew this was going to be issue, I was thinking through something. might be able to test it next commit.

tobgu commented on qframe_test.go:
```diff
@@ -1373,6 +1373,65 @@ mon
 	})
 }
 
+func TestQFrame_ReadCSVMissingColumnName(t *testing.T) {
```
These tests would probably lend themselves nicely to "table driven testing" if that's something you want to explore: https://github.com/golang/go/wiki/TableDrivenTests

kp2401075 replied:
I think I'm going to have to spend some more time getting familiar with TableDrivenTests.

Might do another PR for documentation seprately if that's ok.

tobgu replied:
If you have things that you know you'd like to do I'd prefer if we did it all in this PR to make it as complete. Otherwise I'm OK with the current state of affairs. Examples are great  but not a must.

tobgu commented on internal/io/csv.go:
```diff
@@ -98,6 +101,40 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
+		headersMap := make(map[string]int)
+		// Loop through all columsn and find the duplicates.
+		for i, h := range headers {
+			headersMap[h] = i
```
I personally would prefer if this was the first index of the header rather than the last.
Eg. `COL,COL,COL` -> `COL,COL0,COL1`, right now it's `COL,COL,COL` -> `COL0,COL1,COL`.

tobgu commented on internal/io/csv.go:
```diff
@@ -98,6 +101,40 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
+		for i, h := range headers {
+			index, ok := headersMap[h]
+			if ok && i != index {
+				counter := 0
```
I like this approach better than the previous since it's more straight forward to reason about. I would like to change the name of the config variable though since the new approach has nothing to do with the column indicies.
Perhaps `RenameDuplicateColumns`?

tobgu commented on internal/io/csv.go:
```diff
@@ -98,6 +101,40 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
+			if ok && i != index {
+				counter := 0
+				for {
+					_, ok = headersMap[headers[i]+fmt.Sprint(counter)]
```
Assign `headers[i]+fmt.Sprint(counter)` to a variable (`candidateName`?) since it's used multiple times?

tobgu replied:
Inline and assign `_, ok =` in the if-expression to reduce the scope of the variable and avoid overwriting the existing `ok` variable?

tobgu commented on internal/io/csv.go:
```diff
@@ -98,6 +101,40 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
+	}
+
+	if conf.AddColumnIndexToDuplicateColumns {
+		headersMap := make(map[string]int)
```
Consider making a new function for this code block taking headers and returning new headers since this is now becoming a quite significant piece of code.

tobgu commented on internal/io/csv.go:
```diff
@@ -98,6 +101,40 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
 	}
 
+	if conf.MissingColumnNameAlias != "" {
+		for i, name := range headers {
```
Consider moving to its own function for symmetry with the below suggestion.

tobgu commented on qframe_test.go:
```diff
@@ -1373,6 +1373,65 @@ mon
+}
+
+func TestQFrame_ReadCSVDuplicateColumnNameWithColumnIndexAdded(t *testing.T) {
+	//Corner case 1
```
I'm not sure these are corner cases anymore with the new method.

tobgu commented on config/csv/config.go:
```diff
@@ -33,6 +33,20 @@ func EmptyNull(emptyNull bool) ConfigFunc {
+}
+
+// AddColumnIndexToDuplicateColumns configures if duplicate column names should have the column index appended to the column name to resolve the conflict.
+func AddColumnIndexToDuplicateColumns(AddColumnIndexToDuplicateColumns bool) ConfigFunc {
```
As mentioned elsewhere please rename this configuration option now that the column indices are not part of the solution. `RenameDuplicateColumns`?

tobgu commented on internal/io/csv.go:
```diff
@@ -151,6 +162,48 @@ func resizeColBytes(bytes [][]byte, currentRowCount, sizeHint int) {
+// Handle DuplicateColumnNames
+func AddColumnIndexToDuplicateColumns(headers []string) []string {
+	headersMap := make(map[string]int)
+	// Loop through all columsn and find the duplicates.
```
Consider rewriting this comment to better reflect why this is done or rename `headersMap` to something more telling and remove comment.

tobgu commented on internal/io/csv.go:
```diff
@@ -151,6 +162,48 @@ func resizeColBytes(bytes [][]byte, currentRowCount, sizeHint int) {
+			headersMap[h] = i
+		}
+	}
+	// For go through all if there is duplicate down the line add index
```
Consider rewriting this comment to better reflect why this is done or removing it completely.

tobgu commented on internal/io/csv.go:
```diff
@@ -151,6 +162,48 @@ func resizeColBytes(bytes [][]byte, currentRowCount, sizeHint int) {
 }
 
+// Handle DuplicateColumnNames
+func AddColumnIndexToDuplicateColumns(headers []string) []string {
```
Make function module private by changing name to start with lower case.

tobgu commented on internal/io/csv.go:
```diff
@@ -151,6 +162,48 @@ func resizeColBytes(bytes [][]byte, currentRowCount, sizeHint int) {
+}
+
+// Handle Missing Columnnames
+func AddAliasToMissingColumnNames(headers []string, alias string) []string {
```
Make function module private by changing name to start with lower case.

tobgu commented on qframe_test.go:
```diff
@@ -1373,6 +1373,41 @@ mon
+}
+
+func TestQFrame_ReadCSVDuplicateColumnName(t *testing.T) {
+	input := `COL,COL
```
Please update this input to something like `COL,COL,COL,KOL,KOL` to better capture the essence of the current behavior.

tobgu reviewed (changes requested):
Great! I've added a couple of minor comments. Once those are fixed I think we're good to merge!

tobgu commented on config/csv/config.go:
```diff
@@ -33,6 +33,20 @@ func EmptyNull(emptyNull bool) ConfigFunc {
+	}
+}
+
+// AddColumnIndexToDuplicateColumns configures if duplicate column names should have the column index appended to the column name to resolve the conflict.
```
Update doc string to reflect new name and functionality

tobgu commented on internal/io/csv.go:
```diff
@@ -151,6 +162,49 @@ func resizeColBytes(bytes [][]byte, currentRowCount, sizeHint int) {
 	}
 }
 
+//
```
Remove empty comment

tobgu commented on qframe_test.go:
```diff
@@ -1373,6 +1373,41 @@ mon
+	assertEquals(t, expected, out)
+}
+
+func TestQFrame_ReadCSVDuplicateColumnName(t *testing.T) {
```
This test does not assert anything meaningful ATM, there is an error in the header (a missing comma between col two and three) that cause the read method to fail. The resulting frame has `Err != nil`. Please fix this and also perhaps add a check that there is no error when reading the frame (`assertNotErr(t, out.Err)`) to be on the safe side.

tobgu reviewed (changes requested):
Still some things that need fixing.

Once done please squash all commits into one.

Thanks!

kp2401075 commented:
@tobgu , I've tried to squash the commits. not sure it it worked correctly.

Please. review and see if it's really ready for merge.

tobgu commented on examples_test.go:
```diff
@@ -92,13 +93,45 @@ c,3.0`
 	fmt.Println(f)
 	// Output:
-	// COL1(s) COL2(f)
+	// COL(s) COL0(f)
```
Not sure what happened here but I can take a look at it. Perhaps the examples need to be added to the automated tests.

tobgu reviewed (approved):
Great, thanks for the effort!

tobgu commented:
Thanks a lot!
